notebooks/k_nearest_neighbors.livemd

```diff
@@ -0,0 +1,807 @@
+# K-Nearest Neighbors
+
+```elixir
+Mix.install([
+  {:scholar, github: "elixir-nx/scholar"},
+  {:explorer, "~> 0.5.5"},
+  {:exla, "~> 0.5.2"},
+  {:req, "~> 0.3.6"},
+  {:kino_vega_lite, "~> 0.1.8"},
+  {:kino, "~> 0.9.0"},
+  {:scidata, "~> 0.1.9"},
+  {:kino_explorer, "~> 0.1.2"}
+])
+```
+
+## Setup
+
+We will extensively use VegaLite, Explorer, and Scholar throughout this guide, so let's define some aliases:
+
+```elixir
+alias VegaLite, as: Vl
+require Explorer.DataFrame, as: DF
+require Explorer.Series, as: S
+require Explorer.Query, as: Q
+alias Scholar.Neighbors.KNearestNeighbors
+```
+
+And let's configure `EXLA.Backend` as our default across the notebook and all branched sections:
+
+```elixir
+Nx.global_default_backend(EXLA.Backend)
+```
+
+## Introduction
+
+This notebook will cover the three primary applications of K-Nearest Neighbors: classification, regression, and
+anomaly detection.
+
+We will also use a different approach for this notebook. First, we are going to implement KFold, and then we will explain
+how it works later on. So feel free to skip the implementation below for now and come back later once you grasp the concepts.
+
+```elixir
+defmodule KFold do
```

Comment:
Should we provide KFold as part of Scholar itself? I think scikit-learn ships with its own version too, right?

Reply:
It would be great if we provide this as a part of Scholar, but I think it might be a little bit problematic. Cross-validation goes through a given set of hyperparameters which very often are numbers because tensors' shapes depend on them. In this case, I also encounter the problem, because we iterate through different values of num_neighbors which must be a number. When I tried implementation with pure Nx and while I cannot implement this to pass num_neighbors as a number.

Reply:
@msluszniak I see. I think it is fine then to implement the outer loop in Elixir, as you have done here, as long as the inner of the loop is in Nx. As long as we document it and we don't name the function `fit/predict`, which is not the case in scikit anyway, i think we are fine. :)